webapp/packages/plugin-data-spreadsheet-new/src/DataGrid/DataGridTableContainer.tsx

```diff
@@ -0,0 +1,108 @@
+/*
+ * cloudbeaver - Cloud Database Manager
+ * Copyright (C) 2020 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+
+import { useCallback, useContext } from 'react';
+import styled, { css } from 'reshadow';
+
+import type { SqlResultSet } from '@cloudbeaver/core-sdk';
+import { copyToClipboard } from '@cloudbeaver/core-utils';
+import type { IDatabaseDataResult } from '@cloudbeaver/plugin-data-viewer';
+
+import { DataGridSelectionContext } from './DataGridSelection/DataGridSelectionContext';
+import type { IPosition } from './DataGridSelection/useGridSelectionContext';
+
+const styles = css`
+  grid-container {
+    display: flex;
+    flex-direction: column;
+    flex: 1;
+  }
+`;
+
+interface Props {
+  children: React.ReactNode;
+  modelResultData: IDatabaseDataResult | null;
+}
+
+const EVENT_KEY_CODE = {
+  C: 'KeyC',
+};
+
+function copyGridSelectedDataToClipboard(modelData: SqlResultSet, selectedCells: Map<string, IPosition>) {
+  if (!modelData.rows) {
+    return;
+  }
+
+  const orderedSelectedCells = Array.from((selectedCells.values())).sort((a, b) => {
+    if (a.rowIdx === b.rowIdx) {
+      return a.idx - b.idx;
+    }
+    return a.rowIdx - b.rowIdx;
+  });
+
+  const selectedRows: Map<number, Set<number>> = new Map();
+  const selectedColumns: Set<number> = new Set();
+
+  for (const { idx, rowIdx } of orderedSelectedCells.values()) {
+    if (!selectedColumns.has(idx)) {
+      selectedColumns.add(idx);
+    }
+
+    if (!selectedRows.has(rowIdx)) {
+      selectedRows.set(rowIdx, new Set([idx]));
+      continue;
+    }
+    selectedRows.set(rowIdx, selectedRows.get(rowIdx)!.add(idx));
+  }
+
+  const sortedSelectedColumns = Array.from(selectedColumns.values()).sort();
+
+  let data = '';
+  for (const [rowIdx, colIndexes] of selectedRows.entries()) {
+    for (const column of sortedSelectedColumns) {
+      if (column !== sortedSelectedColumns[0]) {
+        data += '\t';
+      }
+
+      if (colIndexes.has(column)) {
+        const value = modelData.rows?.[rowIdx][column - 1];
+        data += value;
+      }
+    }
+    data += '\r\n';
+  }
+
+  copyToClipboard(data);
```

Comment:
code is really complicated here, you should work with structure like `Map<number, number[]>` where key is the row index, and value is column indexes selected in this row

Reply:
it gives you the opportunity to sort rows fast and the easiest way to calculate selected columns to determine the table's structure